### PR TITLE
Handle changes to election data model.

### PR DESCRIPTION
**Pull request by ericnormand**
There are two places where electorate-works was affected. When an
election create even comes in, we have to grab the right keys and
iterate through its district divisions. For user create event, we need
to look for the correct district-division that matches the user after we
search.

### CODE REVIEW

File: src/electorate_works/handlers.clj [hunk 1]
```diff
@@ -136,26 +136,30 @@
...
        (for [election (-> (search-elections {:district-divisions user-ocdids})
                           (deref-or-timeout timeout "searching elections")
-                          :elections)]
-         (let [election-level (:election-authority-level election)
-               registration-level (:registration-authority-level election)]
+                          :elections)
+             :let [matching-division (first (filter #(contains? user-ocdids
+                                                                (:ocd-id %))
+                                                    (:district-divisions election)))]
+             :when matching-division]
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: I'm still concerned about duplicating matching logic here.

If we included `?id` in the `:find` clause of [this query](https://github.com/democracyworks/election-works/blob/master/src/election_works/entities/election.clj#L78-L87), could we get the ocd-id that matched each election and return that in the search response (and something similar in user-works)?

File: src/electorate_works/handlers.clj [hunk 2]
```diff
@@ -136,26 +136,30 @@
...
                (assoc-when :election-authority
                            (lookup-authority mem-search
                                              timeout :election
-                                             ocds-by-level election-level))
+                                             @ocds-by-level election-level))
                (assoc-when :registration-authority
                            (lookup-authority mem-search
                                              timeout :registration
-                                             ocds-by-level registration-level)))))))))
+                                             @ocds-by-level registration-level)))))))))
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: So were these not working before?

File: src/electorate_works/handlers.clj [hunk 3]
```diff
@@ -183,30 +187,29 @@
...
-               (assoc-when :election-authority @election-authority)
-               (assoc-when :registration-authority @reg-authority))))))))
+  (let [mem-search (memoize search-election-authorities)]
+    (not-empty
+     (doall
+      (for [district-division (:district-divisions election)
+            user (-> (search-users {:district-division (:ocd-id district-division)})
+                     (deref-or-timeout timeout "searching users")
+                     :users)]
+        (let [election-level (:election-authority-level district-division)
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: How about `election-authority-level` to make sure it's clear what this var holds?

File: src/electorate_works/handlers.clj [hunk 4]
```diff
@@ -183,30 +187,29 @@
...
+              registration-level (:voter-registration-authority-level district-division)
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: And `registration-authority-level`?

File: src/electorate_works/handlers.clj (hunk 2 above)
Comment (ericnormand) [inline src/electorate_works/handlers.clj]: Doubtful. There were no tests and `nil` was allowed.

File: src/electorate_works/handlers.clj (hunk 1 above)
Comment (ericnormand) [inline src/electorate_works/handlers.clj]: That's certainly possible. 

BUT!!

Where do we put that data in the electorate? We pass along the user and the election, with no information about how they were matched.

Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Do we need it in the electorate? I was just thinking we could use it to determine how to get the 2 authorities, associate them in the electorate, and be done.

Then if we change the matching logic in election-works or user-works, things will keep working.

Comment (ericnormand) [inline src/electorate_works/handlers.clj]: But what about determining deadlines?

Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Ohhhh... Good point. Yeah, we would need to store the matching OCD-ID somewhere in the electorate to determine the right district-division to use for that. But that doesn't seem too difficult, does it? We could just add an attribute for that and send it along in the events.

Comment (ericnormand) [inline src/electorate_works/handlers.clj]: This might be too big of a change to decide in a PR.

Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Yeah, since this isn't pressing at the moment, let's write cards for what we think should happen from here.

Comment (ericnormand) [inline src/electorate_works/handlers.clj]: Will do!

**Comment by cap10morgan** [comment]
With the cards to address the post-beta cleanup we'd like to do around not duplicating matching logic among three different components, this LGTM for now.

**Comment by ericnormand** [comment]
Ok. Waiting for a coordinated merge/deploy with election-works.

**Comment by ericnormand** [comment]
Ok, I've got it working with election-works in Krakenstein. I can do a demo when you get back, @cap10morgan .

File: test/electorate_works/handlers_test.clj [hunk 5]
```diff
@@ -136,10 +136,27 @@
...
 
+(def user-with-la-ocd
+  {:id (java.util.UUID/randomUUID)
+   :addresses {:registered {:ocd-divisions #{"ocd-division/country:us/state:la"}}}})
+
 (defn ref-one-election [& args]
   (future
     {:elections #{{:id (java.util.UUID/randomUUID)
-                   :district-divisions #{"ocd-division/country:us"}}}}))
+                   :district-divisions #{{:ocd-id "ocd-division/country:us"}}}}}))
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: This won't ever be the district division of a real election in the current data model. Maybe we should test with state or lower OCD-IDs?

File: test/electorate_works/handlers_test.clj [hunk 6]
```diff
@@ -136,10 +136,27 @@
...
+
+(def election-with-la-ocd
+  {:id (java.util.UUID/randomUUID)
+   :district-divisions #{{:ocd-id "ocd-division/country:us/state:la"
+                          :election-authority-level :state
+                          :voter-registration-authority-level :state}}})
+
+(defn ref-la-election [{:keys [district-divisions]}]
+  (clojure.pprint/pprint district-divisions)
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: Should this still be in here?

**Comment by cap10morgan** [comment]
LGTM

**Pull request merged**